### PR TITLE
Migrate Hive connector to new table layout API

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import static com.facebook.presto.hive.util.Types.checkType;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Predicates.not;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.String.format;
+import static org.apache.hadoop.hive.metastore.ProtectMode.getProtectModeFromString;
+
+public class HivePartitionManager
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java]: Almost exact copy of the functions removed from HiveSplitManager

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionResult.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Result of fetching Partitions in the HivePartitionManager interface.
+ *
+ * Results are comprised of two parts:
+ * 1) The actual partitions
+ * 2) The TupleDomain that represents the values that the connector was not able to pre-evaluate
+ * when generating the partitions and will need to be double-checked by the final execution plan.
+ */
+public class HivePartitionResult
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionResult.java]: Almost exact copy of ConnectorPartitionResult

File: presto-hive/src/main/java/com/facebook/presto/hive/util/ImmutableCollectors.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultiset;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+public final class ImmutableCollectors
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/util/ImmutableCollectors.java]: Exact copy from presto-main `com.facebook.presto.util`

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/util/ImmutableCollectors.java]: This class isn't needed. Just use the regular collectors in the JDK. We will add `ImmutableCollectors` to Airlift at some point and can switch over then.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java (hunk 1 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java]: I understand the motivation for this refactoring, but it should be done separately from switching to the layout API. They don't seem to be related.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionResult.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+{
+    private final List<HivePartition> partitions;
+    private final TupleDomain<ColumnHandle> unenforcedTupleDomain;
```
Comment (martint) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionResult.java]: I'd call this `unenforcedConstraint`

**Comment by martint** [comment]
Other than David's comments, it looks good

**Comment by haozhun** [comment]
@martint  Do you want to take another look before merge?

**Comment by martint** [comment]
looks good

**Comment by haozhun** [comment]
Merged #3475
